tidy3d/components/source.py

```diff
@@ -509,6 +500,114 @@ def _warn_if_large_number_of_freqs(cls, val):
+    or the source plane must span the entire simulation domain and the fields must match the
+    simulation boundary conditions. The equivalent source currents are fully defined by the field
+    components tangential to the source plane. The normal components (e.g. ``Ez`` and ``Hz``) can be
+    provided but will have no effect on the results, in accordance with the equivalence principle.
+    At least one of the tangential components has to be defined. For example, for a ``z``-normal
+    source, at least one of ``Ex``, ``Ey``, ``Hx``, and ``Hy`` has to be present in the provided
+    dataset. The coordinates of all provided fields are assumed to be relative to the source
+    center. Each provided field component must also span the size of the source.
+
+    Note
+    ----
+        If only the ``E`` or only the ``H`` fields are provided, the source will not be directional,
+        but will inject equal power in both directions instead.
+
+    Example
+    -------
+    >>> from tidy3d import ScalarFieldDataArray
+    >>> pulse = GaussianPulse(freq0=200e12, fwidth=20e12)
+    >>> x = np.linspace(-1, 1, 101)
+    >>> y = np.linspace(-1, 1, 101)
+    >>> z = np.array([0])
+    >>> f = [2e14]
+    >>> coords = dict(x=x, y=y, z=z, f=f)
+    >>> scalar_field = ScalarFieldDataArray(np.ones((101, 101, 1, 1)), coords=coords)
+    >>> dataset = FieldDataset(Ex=scalar_field)
+    >>> custom_source = CustomFieldSource(
+    ...     center=(1, 1, 1),
+    ...     size=(2, 2, 0),
+    ...     source_time=pulse,
+    ...     field_dataset=dataset)
+
+    """
+
+    field_dataset: Optional[FieldDataset] = pydantic.Field(
+        ...,
+        title="Field Dataset",
+        description=":class:`.FieldDataset` containing the desired frequency-domain "
+        "fields patterns to inject. At least one tangetial field component must be specified.",
+    )
+
+    @pydantic.validator("field_dataset", pre=True, always=True)
+    def _warn_if_none(cls, val: FieldDataset) -> FieldDataset:
+        """Warn if the DataArrays fail to load."""
+        if isinstance(val, dict):
+            if DATA_ARRAY_TAG in [v for v in val.values() if isinstance(v, str)]:
+                log.warning("Loading 'field_dataset' without data.")
+                return None
+        return val
+
+    @pydantic.validator("field_dataset", always=True)
+    def _single_frequency_in_range(cls, val: FieldDataset, values: dict) -> FieldDataset:
+        """Assert only one frequency supplied and it's in source time range."""
+        if val is None:
+            return val
+        source_time = get_value(key="source_time", values=values)
+        fmin, fmax = source_time.frequency_range()
+        for name, scalar_field in val.field_components.items():
+            freqs = scalar_field.f
+            if len(freqs) != 1:
+                raise SetupError(
+                    f"`field_data.{name}` must have a single frequency, "
```

Comment:
fix single quotes in these raise statements, (this and below) (sorry about that)